### DISCUSSION TITLE
facing problem while implementing plugin development

### DISCUSSION THREAD

**Discussion by ThisIsFaar** [joomla/joomla-cms#34556] [Q&A]
I was learning plugin development from Joomla docs, I want to add my js file in the head by a plugin that I am working on

I am using onBeforeCompileHead event.

**PROBLEM:** not able to add my js file by this code, also implement a couple of ways nothing works for me. 

Code is below

```
<?php
defined('_JEXEC') or die;

use Joomla\CMS\Plugin\CMSPlugin;
use Joomla\CMS\Factory;
$document = Factory::getDocument();
use Joomla\CMS\Uri\Uri;
use Joomla\Event\Event;
use Joomla\Event\SubscriberInterface;


class plgSystemInjectName extends CMSPlugin implements SubscriberInterface
{
    /**
	 * Returns an array of events this subscriber will listen to.
	 *
	 * @return  array
	 */
	public static function getSubscribedEvents(): array
	{
		return [
			'onBeforeCompileHead' => 'myFunctionName',
		];
	}

	/**
	 * Plugin method is the array value in the getSubscribedEvents method
	 * The plugin then modifies the Event object (if it's not immutable)
	 */
	 public function myFunctionName(Event $event)
	 {
		/*
		 * Plugin code goes here.
		 * You can access parameters via $this->params
		 */
        $url = Uri::base().'plugins/system/injectname/js/main.js';
        $document->addScript($url);


		return true;
	}
}
```

**Comment by Fedik** [comment]
Never call `Factory::getDocument();` as you doing, it should be called not earlier than onAfterDispatch event.
Also it is recomended that you store all asset files (js, css) under /media folder. 
Here example how to add media files in to a manifest of your plugin https://docs.joomla.org/Manifest_files#Media_files

Kind of this:
```xml
<media folder="media" destination="plg_system_injectname">
	<folder>js</folder>
</media>
```

For joomla 4 look https://docs.joomla.org/J4.x:Web_Assets how to attach asset to document.
Basically inside `onBeforeCompileHead` event you do something like:
```php
/** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
$wa = Factory::getApplication()->getDocument()->getWebAssetManager();
$wa->registerAndUseScript('plg_system_injectname', 'plg_system_injectname/main.js', []);
```
`plg_system_injectname/main.js` - assuming that the file in `/media/plg_system_injectname/js/main.js`

**Reply by ThisIsFaar** [reply]
> Never call `Factory::getDocument();` as you doing, it should be called not earlier than onAfterDispatch event.
> Also it is recomended that you store all asset files (js, css) under /media folder.
> Here example how to add media files in to a manifest of your plugin https://docs.joomla.org/Manifest_files#Media_files
> 
> Kind of this:
> 
> ```
> <media folder="media" destination="plg_system_injectname">
> 	<folder>js</folder>
> </media>
> ```
> 
> For joomla 4 look https://docs.joomla.org/J4.x:Web_Assets how to attach asset to document.
> Basically inside `onBeforeCompileHead` event you do something like:
> 
> ```
> /** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
> $wa = Factory::getApplication()->getDocument()->getWebAssetManager();
> $wa->registerAndUseScript('plg_system_injectname', 'plg_system_injectname/main.js', []);
> ```
> 
> `plg_system_injectname/main.js` - assuming that the file in `/media/plg_system_injectname/js/main.js`

forget the previous code,

now I just want to add inline js below is the code that is not working: 


```
<?php
defined('_JEXEC') or die;

use Joomla\CMS\Plugin\CMSPlugin;
use Joomla\CMS\Factory;

class plgSystemInjectName extends CMSPlugin
{
    public function onAfterDispatch () {
        $document = Factory::getDocument();
        $document->addScriptDeclaration('
            window.event("domready", function() {
                alert("An inline JavaScript Declaration");
            });
        ');
    }
}
```

**Reply by Fedik** [reply]
because it should not :)

joomla 3:
```php
$document = Factory::getDocument();
$document->addScriptDeclaration('
   alert("An inline JavaScript Declaration");
');
```

joomla 4:
```php
$wa = Factory::getApplication()->getDocument()->getWebAssetManager();
$wa->addInlineScript('
  alert("An inline JavaScript Declaration");
');
```
